### ISSUE TITLE
Newick tree writer for hclust

### ISSUE THREAD

**Issue by TomConlin** [JuliaStats/Clustering.jl#209]
More a question than issue.

After generating a cluster, needed to write it out in a portable format.
Not finding one handy, I ported one from R.
Is this something I should consider offering here 
or is tree serializiation out of scope?

**Comment by alyst** [comment]
> Is this something I should consider offering here or is tree serializiation out of scope?

Thank you for considering this contribution!
Unfortunately, I haven't checked the current state of the hierarchical structures ecosystem in Julia, maybe there's more appropriate package for it.
It's true that serialization is somewhat marginal to the clustering package.
But if hclust serialization code is compact and doesn't bring any new dependencies, it's reasonable to include it.
(*Clustering* is sort of a *basic* package, and we want to keep basic package dependencies at minimum.)

Does it come with deserialization support as well? Would it be possible to read the files saved by the other software?

**Comment by TomConlin** [comment]
I took a quick pass through Julia observer before asking here.
It all hinges on others using your same hclust struct.

This is maybe my third dabbing in Julia, previous were at 
v0.2 & v0.7 so I won't be clued into the current ecosystem.

The script is a straight port of  
    https://rdrr.io/bioc/ctc/src/R/hc2Newick.R
so zero dependencies. 

I very much agree scope creep is not to be taken on without
good and thorough consideration. 
I also see how being able to convert Newick trees found in the wild back into distance matricies could be generally useful ... it is good; but nerd sniping me will have to wait on real work for now.

This function just allows other Newick tree readers to read/display
trees generated by Clustering.jl.

**Comment by alyst** [comment]
> The script is a straight port of https://rdrr.io/bioc/ctc/src/R/hc2Newick.R

Ah, that looks like something that could be put into Clustering.jl!

Just when porting it to Julia, please rewrite it so that it uses the IO for output (i.e. the function should look `write_newick(io::IO, hclu::Hclust)`).
Then it would be possible to write hclu directly into a file. It would be also very easy to pass `io::IOBuffer` into `write_newick()` and then `String(take!(io))`, to get a string, but that would be much more efficient than string concatenation.
To merge the PR would we need that the function is tested. The good test would be to check that for a known tree(s) the output matches the saved example.

**Comment by TomConlin** [comment]
Thanks for the encouragement. I will spend a bit of time trying to absorb enough Julia idioms and move my implementation from something that works to something that won't make a more experienced developers eyes bleed.

**Comment by TomConlin** [comment]
Have some working parts for this at
https://github.com/TomConlin/Clustering.jl/tree/Newick_Tree
not PR ready as it should get some reorganization of 
what is kept and where it should go

**Comment by arzwa** [comment]
Hi, I stumbled on this issue by chance, I happen to have needed this too at one point. I wrote the following function using a small module for reading/writing trees I wrote for my phylogenetics work, might be useful:

```julia
using NewickTree, Clustering
using NewickTree: setdistance!

function get_tree(hc)
   nodes = [Node(i, n=string(n), d=0.) for (i,n) in zip(hc.order,hc.labels)]
   n = length(nodes)
   idfun(x) = x > 0 ? x + n : abs(x)
   for i=1:size(hc.merges, 1)
       nid = n + i
       j, k = idfun.(hc.merges[i,:])
       a = nodes[j]
       b = nodes[k]
       h = hc.heights[i]
       newnode = Node(nid, n="$nid", d=h)
       setdistance!(a, h-distance(a))
       setdistance!(b, h-distance(b))
       push!(newnode, a)
       push!(newnode, b)
       push!(nodes, newnode)
   end
   setdistance!(nodes[end], 0.)
   return nodes[end]
end

D = rand(10, 10)
D += D'
hc = hclust(D)
tr = get_tree(hc) 
```

**Comment by TomConlin** [comment]
@alyst I think closing this and letting it point to @arzwa  solution is optimal.
arzwa  you are welcome to salvage any bits from my R adaption, 
a test maybe is all that is useful.
